pkgs/development/libraries/protobuf/generic-v3.nix

```diff
@@ -1,60 +1,68 @@
-{ stdenv
-, fetchFromGitHub
-, autoreconfHook, zlib, gmock, buildPackages
+{ stdenv, fetchFromGitHub, lib
+, zlib, cmake, gmock
 , version, sha256
 , ...
 }:
-
-let
-mkProtobufDerivation = buildProtobuf: stdenv: stdenv.mkDerivation {
+stdenv.mkDerivation rec {
+  # make sure you test also -A pythonPackages.protobuf
   pname = "protobuf";
   inherit version;
-
-  # make sure you test also -A pythonPackages.protobuf
   src = fetchFromGitHub {
     owner = "protocolbuffers";
     repo = "protobuf";
     rev = "v${version}";
     inherit sha256;
   };
 
-  postPatch = ''
-    rm -rf gmock
-    cp -r ${gmock.src}/googlemock gmock
-    cp -r ${gmock.src}/googletest googletest
-    chmod -R a+w gmock
-    chmod -R a+w googletest
-    ln -s ../googletest gmock/gtest
-  '' + stdenv.lib.optionalString stdenv.isDarwin ''
-    substituteInPlace src/google/protobuf/testing/googletest.cc \
-      --replace 'tmpnam(b)' '"'$TMPDIR'/foo"'
-  '';
+  patches =
+    [ ./protobuf-cmake.patch ]
+    # In version 3.6.X,
+    # CMake doesn't generate the version.rc file from version.rc.in.
+    ++ lib.optional
+       (lib.versionAtLeast version "3.6" && lib.versionOlder version "3.7")
+       ./protobuf-3.6-cmake-versionrc.patch;
 
-  nativeBuildInputs = [ autoreconfHook buildPackages.which buildPackages.stdenv.cc buildProtobuf ];
+  postPatch = 
+    (if lib.versionAtLeast version "3.6.0"
+     then ''
+       rm -rf third_party/googletest
+       cp -a ${gmock.src} third_party/googletest
+       chmod -R a+w third_party/googletest
+       ''
+     else ''
+     cp -a ${gmock.src}/googlemock gmock
+     cp -a ${gmock.src}/googletest googletest
+     chmod -R a+w gmock
+     chmod -R a+w googletest
+     ln -s ../googletest gmock/gtest
+     '')
+    + lib.optionalString stdenv.isDarwin ''
+      substituteInPlace src/google/protobuf/testing/googletest.cc \
+      --replace 'tmpnam(b)' '"'$TMPDIR'/foo"'
+      '';
 
+  nativeBuildInputs = [ cmake ];
   buildInputs = [ zlib ];
-  configureFlags = if buildProtobuf == null then [] else [ "--with-protoc=${buildProtobuf}/bin/protoc" ];
```

Comment:
Did you check that cross builds still function?

Reply:
I believe so. protobuf builds protoc but doesn't actually use it as part of the build, it appears someone misread the instructions for cross compiling projects that depend on protobuf as directions for cross-compiling protobuf. But I'll run the build one more time to be sure.

Reply:
`nix-build -A pkgsCross.armv7l-hf-multiplatform.protobuf` is now failing with a `not found` rather than a `exec format error` for protoc (the latter would have been what I expected. I'll investigate in a few hours.